### PR TITLE
Fixes for Windows issues, and add non-linux os CI tests

### LINKED ISSUE #411: RuntimeError: expected scalar type Long but found Int in [cd_spot_the_diff_mnist_wine.ipynb]
I am getting the following error when trying to execute code (in [10] section "Interpretable Drift Detection on the Wine Quality Dataset"):
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_3564/2414863533.py in <module>
     10 )
     11 
---> 12 preds_h0 = cd.predict(x_h0)
     13 preds_corr = cd.predict(x_corr)

~\AppData\Roaming\Python\Python39\site-packages\alibi_detect\cd\spot_the_diff.py in predict(self, x, return_p_val, return_distance, return_probs, return_model)
    173         data, and the trained model.
    174         """
--> 175         return self._detector.predict(x, return_p_val, return_distance, return_probs, return_model)

~\AppData\Roaming\Python\Python39\site-packages\alibi_detect\cd\pytorch\spot_the_diff.py in predict(self, x, return_p_val, return_distance, return_probs, return_model)
    212         data, and the trained model.
    213         """
--> 214         preds = self._detector.predict(x, return_p_val, return_distance, return_probs, return_model=True)
    215         preds['data']['diffs'] = preds['data']['model'].diffs.detach().cpu().numpy()  # type: ignore
    216         preds['data']['diff_coeffs'] = preds['data']['model'].coeffs.detach().cpu().numpy()  # type: ignore

~\AppData\Roaming\Python\Python39\site-packages\alibi_detect\cd\base.py in predict(self, x, return_p_val, return_distance, return_probs, return_model)
    241         """
    242         # compute drift scores
--> 243         p_val, dist, probs_ref, probs_test = self.score(x)
    244         drift_pred = int(p_val < self.p_val)
    245 

~\AppData\Roaming\Python\Python39\site-packages\alibi_detect\cd\pytorch\classifier.py in score(self, x)
    182             self.model = self.model.to(self.device)
    183             train_args = [self.model, self.loss_fn, dl_tr, self.device]
--> 184             trainer(*train_args, **self.train_kwargs)  # type: ignore
    185             preds = self.predict_fn(x_te, self.model.eval())
    186             preds_oof_list.append(preds)

~\AppData\Roaming\Python\Python39\site-packages\alibi_detect\models\pytorch\trainer.py in trainer(model, loss_fn, dataloader, device, optimizer, learning_rate, preprocess_fn, epochs, reg_loss_fn, verbose)
     53             y_hat = model(x)
     54             optimizer.zero_grad()  # type: ignore
---> 55             loss = loss_fn(y_hat, y) + reg_loss_fn(model)
     56             loss.backward()
     57             optimizer.step()  # type: ignore

~\anaconda3\lib\site-packages\torch\nn\modules\module.py in _call_impl(self, *input, **kwargs)
   1100         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1101                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1102             return forward_call(*input, **kwargs)
   1103         # Do not call functions when jit is used
   1104         full_backward_hooks, non_full_backward_hooks = [], []

~\anaconda3\lib\site-packages\torch\nn\modules\loss.py in forward(self, input, target)
   1148 
   1149     def forward(self, input: Tensor, target: Tensor) -> Tensor:
-> 1150         return F.cross_entropy(input, target, weight=self.weight,
   1151                                ignore_index=self.ignore_index, reduction=self.reduction,
   1152                                label_smoothing=self.label_smoothing)

~\anaconda3\lib\site-packages\torch\nn\functional.py in cross_entropy(input, target, weight, size_average, ignore_index, reduce, reduction, label_smoothing)
   2844     if size_average is not None or reduce is not None:
   2845         reduction = _Reduction.legacy_get_string(size_average, reduce)
-> 2846     return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
   2847 
   2848 

RuntimeError: expected scalar type Long but found Int
----
I use Python 3.8.8/ Win10 installed on the AMD Ryzen with integrated graphics (AMD).

### PR DESCRIPTION
**Pull request by ascillitoe**
This PR fixes a number of issues encountered when running on WIndows. It also adds CI build tests for non-linux OS', namely Windows and MacOS. To reduce compute the non-linux OS' are only tested on Python 3.9. 

The notebook tests are also extended to the non-linux OS', mainly to add coverage for the model/dataset fetch functions (#421 shows some of these are currently failing on Windows). 

Fixes #411 and #421.

TODO:
- [x] Get build CI running.
- [x] Extend to notebook tests.
- [x] Fix any newly exposed errors (running tests on all notebooks might expose additional errors on top of those already fixed).

### CODE REVIEW

**Comment by ascillitoe** [comment]
### Results
Results on Windows: 56 failed, 1806 passed, 10 skipped

Issues to fix are:
- [ ] All `test_clfdrift` tests fail. Looks like #411.
- [ ] All `test_stddrift` tests fail. Again, looks like #411 (expected as `SpotTheDiffDrift` is based on `ClassifierDrift`).

**Comment by ascillitoe** [comment]
@jklaise what are your thoughts on notebooks, shall we add those too? I'm undecided, I'm not sure we should really need to if our test coverage is good enough, but as #421 demonstrates we do currently have a few platform-specific issues in the notebooks themselves (i.e. `os.path` usage), so maybe it wouldn't hurt to add for now?

One minor issue I foresee will be we would have to exclude `od_prophet_weather.ipynb` from Windows tests for now. I've excluded `OutlierProphet` from Windows tests as `pystan` (`fbprophet` dependency) is a bit more challenging to install on Windows.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -17,39 +21,55 @@ jobs:
...
+
       - name: Install dependencies
         run: |
-          sudo apt-get install build-essential
-          pip install --upgrade pip
-          pip install -r requirements/dev.txt
-          pip install -e .[prophet]
-          pip install -e .[torch]
-          pip freeze
+          # sudo apt-get install build-essential # build-essential appears to already be up-to date on all latest github os'. Nice to remove as otherwise need conditional step dependent on os
```
Comment (ascillitoe) [inline .github/workflows/ci.yml:46]: With the ubuntu builds this line was outputting:
```
0 upgraded, 0 newly installed, 0 to remove and 17 not upgraded.
```

so it looks like it isn't currently needed? This line breaks the windows and macos builds (as don't have `apt-get`), so removing is nice as avoids the need for `if: matrix.os == ...` logic here.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -17,39 +21,55 @@ jobs:
...
+          python -m pip install --upgrade pip
+          python -m pip install wheel # To avoid having to install dependencies from setup.py's
+          python -m pip install -r requirements/dev.txt
+          if [ "$RUNNER_OS" != "Windows" ]; then  # Skip Prophet tests on windows for now (need to ensure mingw-w64 compiler toolchain installed properly)
```
Comment (ascillitoe) [inline .github/workflows/ci.yml:50]: `OutlierProphet` tests are skipped on Windows for now as it was proving difficult to install cleanly (due to need to install `pystan`).

File: alibi_detect/cd/base.py [hunk 3]
```diff
@@ -151,7 +151,7 @@ def get_splits(self, x_ref: Union[np.ndarray, list], x: Union[np.ndarray, list])
         train and test indices for optionally different folds.
         """
         # create dataset and labels
-        y = np.concatenate([np.zeros(len(x_ref)), np.ones(len(x))], axis=0).astype(int)
+        y = np.concatenate([np.zeros(len(x_ref)), np.ones(len(x))], axis=0).astype(np.int64)  # Fix #411
```
Comment (ascillitoe) [inline alibi_detect/cd/base.py:154]: Fix for #411

**Comment by ascillitoe** [comment]
@jklaise this is ready for review. I've tested the updated fetch methods on a few random notebooks and they seem to work, but I didn't cover all possible fetch calls. We can either merge this PR into a new branch (so that we can manually trigger the `test_all_notebooks`), or merge into master and deal with any outstanding notebook issues in a later PR. (By outstanding issues I mean newly discovered Windows issues, and any possibly introduced by the updated fetch methods).

File: .github/workflows/test_all_notebooks.yml [hunk 4]
```diff
@@ -27,12 +37,13 @@ jobs:
...
-          pip install --upgrade pip
-          pip install -r requirements/dev.txt
-          pip install -e .[prophet]
-          pip install -e .[torch]
-          pip freeze
+          python -m pip install --upgrade pip
+          python -m pip install -r requirements/dev.txt
+          if [ "$RUNNER_OS" != "Windows" ]; then  # Skip Prophet notebooks on windows for now
+            python -m pip install -e .[prophet]
+          fi
```
Comment (jklaise) [inline .github/workflows/test_all_notebooks.yml:44]: I think this won't be as straightforward as the Prophet notebook won't be on the exclusion list, hence will result in a fail.

File: .github/workflows/test_changed_notebooks.yml [hunk 5]
```diff
@@ -43,12 +53,13 @@ jobs:
...
-          pip install --upgrade pip
-          pip install -r requirements/dev.txt
-          pip install -e .[prophet]
-          pip install -e .[torch]
-          pip freeze
+          python -m pip install --upgrade pip
+          python -m pip install -r requirements/dev.txt
+          if [ "$RUNNER_OS" != "Windows" ]; then  # Skip Prophet notebooks on windows for now
+            python -m pip install -e .[prophet]
+          fi
```
Comment (jklaise) [inline .github/workflows/test_changed_notebooks.yml:60]: Same comment as above re notebooks.

File: alibi_detect/datasets.py [hunk 6]
```diff
@@ -21,6 +21,8 @@
 
 logger = logging.getLogger(__name__)
 
+TIMEOUT = 10
```
Comment (jklaise) [inline alibi_detect/datasets.py:24]: Useful to add a short docstring for what this constant is.

File: .github/workflows/test_all_notebooks.yml (hunk 4 above)
Comment (jklaise) [reply .github/workflows/test_all_notebooks.yml:44]: I guess we can use the `PROPHET_INSTALLED` constant.

Comment (ascillitoe) [reply .github/workflows/test_all_notebooks.yml:44]: Ah I did this in `testing/test_notebooks.py`

```python
if PROPHET_INSTALLED:
    EXCLUDE_NOTEBOOKS.add('od_prophet_weather.ipynb')  # Exclude if fbprophet not installed i.e. on Windows
```

Comment (jklaise) [reply .github/workflows/test_all_notebooks.yml:44]: Alright, I need to learn to read to the end of the changes :)

File: alibi_detect/utils/fetching.py [hunk 7]
```diff
@@ -36,6 +37,8 @@
     SpectralResidual
 ]
 
+TIMEOUT = 10
```
Comment (jklaise) [inline alibi_detect/utils/fetching.py:40]: Same short docstring as in `datasets.py`.

File: alibi_detect/utils/fetching.py [hunk 8]
```diff
@@ -511,3 +514,26 @@ def fetch_detector(filepath: Union[str, os.PathLike],
             kwargs = get_pixelcnn_default_kwargs()
     detector = load_detector(filepath, **kwargs)
     return detector
+
+
+def _join_url(base: str, parts: Union[str, List[str]]) -> str:
```
Comment (jklaise) [inline alibi_detect/utils/fetching.py:519]: Should we test this?

Comment (jklaise) [reply alibi_detect/utils/fetching.py:519]: Happy to leave as a TODO to get this merged asap, but the testing functionality wouldn't need to be very extensive as this is just for internal usage.

**Comment by jklaise** [comment]
Nice work Ashley, happy to merge this and we can trigger notebook testing afterwards.

**Review by jklaise — APPROVED** [review]

**Pull request merged**